File: src/main/java/org/hwabeag/divesystem/DiveSystem.java
package org.hwabeag.divesystem;

import org.bukkit.Bukkit;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.plugin.java.JavaPlugin;
import org.hwabeag.divesystem.commands.AmountCommand;
import org.hwabeag.divesystem.commands.DiveCommand;
import org.hwabeag.divesystem.commands.ShopCommand;
import org.hwabeag.divesystem.config.ConfigManager;
import org.hwabeag.divesystem.events.InvClickEvent;
import org.hwabeag.divesystem.events.InvCloseEvent;
import org.hwabeag.divesystem.events.JoinEvent;
import org.hwabeag.divesystem.events.MoveEvent;
import org.hwabeag.divesystem.expansions.DiveSystemExpansion;
import org.hwabeag.divesystem.schedules.DiveCheckTask;

import java.util.Objects;

public final class DiveSystem extends JavaPlugin {

    private static ConfigManager configManager;

    private FileConfiguration config;

    public static DiveSystem getPlugin() {
        return JavaPlugin.getPlugin(DiveSystem.class);
    }

    public static void getConfigManager() {
        if (configManager == null)
            configManager = new ConfigManager();
    }

    private void registerEvents() {
        getServer().getPluginManager().registerEvents(new InvClickEvent(), this);
        getServer().getPluginManager().registerEvents(new InvCloseEvent(), this);
        getServer().getPluginManager().registerEvents(new JoinEvent(), this);
        getServer().getPluginManager().registerEvents(new MoveEvent(), this);
    }

    private void registerCommands() {
        Objects.requireNonNull(getServer().getPluginCommand("잠수금액")).setExecutor(new AmountCommand());
        Objects.requireNonNull(getServer().getPluginCommand("잠수포인트")).setExecutor(new DiveCommand());
        Objects.requireNonNull(getServer().getPluginCommand("잠수상점")).setExecutor(new ShopCommand());
    }

    @Override
    public void onEnable() {
        // Plugin startup logic
        Bukkit.getLogger().info("[DiveSystem] Enable");
        getConfigManager();
        registerCommands();
        registerEvents();
        Bukkit.getScheduler().scheduleAsyncRepeatingTask(this, new DiveCheckTask(), 20*20, 20*20);
        if (Bukkit.getPluginManager().getPlugin("PlaceholderAPI") != null) {
            new DiveSystemExpansion(this).register();
        }
    }

    @Override
    public void onDisable() {
        // Plugin shutdown logic
        Bukkit.getLogger().info("[DiveSystem] Disable");
        ConfigManager.saveConfigs();
    }
}

File: src/main/kotlin/org/hwabeag/divesystem/DiveSystem.kt
package org.hwabeag.divesystem

import org.bukkit.Bukkit
import org.bukkit.plugin.java.JavaPlugin
import org.hwabeag.divesystem.commands.AmountCommand
import org.hwabeag.divesystem.commands.DiveCommand
import org.hwabeag.divesystem.commands.ShopCommand
import org.hwabeag.divesystem.config.ConfigManager
import org.hwabeag.divesystem.data.DatabaseManager
import org.hwabeag.divesystem.events.InvClickEvent
import org.hwabeag.divesystem.events.InvCloseEvent
import org.hwabeag.divesystem.events.JoinEvent
import org.hwabeag.divesystem.events.MoveEvent
import org.hwabeag.divesystem.expansions.DiveSystemExpansion
import org.hwabeag.divesystem.schedules.DiveCheckTask

class DiveSystem : JavaPlugin() {
    companion object {
        lateinit var instance: DiveSystem
            private set
    }

    override fun onEnable() {
        instance = this
        logger.info("[DiveSystem] Enable")

        ConfigManager.initialize(this)
        DatabaseManager.initialize(this)

        registerCommands()
        registerEvents()

        Bukkit.getScheduler().runTaskTimerAsynchronously(this, DiveCheckTask(), 20L * 20, 20L * 20)

        if (Bukkit.getPluginManager().getPlugin("PlaceholderAPI") != null) {
            DiveSystemExpansion(this).register()
        }
    }

    override fun onDisable() {
        logger.info("[DiveSystem] Disable")
        ConfigManager.saveConfigs()
        DatabaseManager.shutdown()
    }

    private fun registerEvents() {
        val pm = server.pluginManager
        pm.registerEvents(InvClickEvent(), this)
        pm.registerEvents(InvCloseEvent(), this)
        pm.registerEvents(JoinEvent(), this)
        pm.registerEvents(MoveEvent(), this)
    }

    private fun registerCommands() {
        val diveCommand = DiveCommand()
        val shopCommand = ShopCommand()
        val amountCommand = AmountCommand()

        getCommand("잠수포인트")?.apply {
            setExecutor(diveCommand)
            tabCompleter = diveCommand
        }
        getCommand("잠수상점")?.apply {
            setExecutor(shopCommand)
            tabCompleter = shopCommand
        }
        getCommand("잠수금액")?.setExecutor(amountCommand)
    }
}
